workers/scheduler/scheduler.go

```diff
@@ -260,6 +261,30 @@ func (s *Scheduler) schedule() {
 	}
 }
 
+// StopPipelineRun will prematurely cancel a pipeline run by killing all of its
+// jobs and running processes immediately.
+func (s *Scheduler) StopPipelineRun(p *gaia.Pipeline, runID int) error {
+
+	// 1. Get all running Jobs
+	// 2. Set state to failed and send a finish signal
+	// 3. Store the result
+
+	pr, err := s.storeService.PipelineGetRunByPipelineIDAndID(p.ID, runID)
+	if err != nil {
+		return err
+	}
+	if pr.Status != gaia.RunRunning {
+		return errors.New("pipeline is not in running state")
+	}
+	for _, job := range pr.Jobs {
+		if job.Status == gaia.JobRunning || job.Status == gaia.JobWaitingExec {
+			job.Status = gaia.JobFailed
+			job.FailPipeline = true
```

Comment:
This is not enough to fail a job. I need to find a way to make jobs that hang cancelled as well. Even if the process is stuck.